### PR TITLE
SQL Project - deploy to docker publish option

### PR DESCRIPTION
**Pull request by llali**
Removed the "Deploy" menu item and instead added "Publish to docker" an option after selecting publish. Now when selecting publish, user see this as options:

Screen Shot 2021-09-10 at 2 48 45 PM

Screen Shot 2021-09-10 at 2 49 17 PM

If user selects" publish to existing server" publish works as before; 
Screen Shot 2021-09-10 at 2 54 54 PM

otherwise, the menu for publish to docker will be shown to use
Screen Shot 2021-09-10 at 2 54 35 PM

Also made password for deploy to docker mandatory. the password won't get generated anymore and user has to enter the value. 

Removed the Azure Function integration with deploy for now. Will have to add a new command to do the publish and app integration

### CODE REVIEW

File: extensions/sql-database-projects/package.json [hunk 1]
```diff
@@ -400,7 +386,6 @@
     "@microsoft/ads-extension-telemetry": "^1.1.5",
     "fast-glob": "^3.1.0",
     "fs-extra": "^5.0.0",
-    "generate-password": "^1.6.0",
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/package.json:403]: update yarn.lock

File: extensions/sql-database-projects/src/controllers/projectController.ts [hunk 2]
```diff
@@ -259,10 +259,10 @@ export class ProjectsController {
 	 * Deploys a project
 	 * @param treeNode a treeItem in a project's hierarchy, to be used to obtain a Project
 	 */
-	public async deployProject(context: Project | dataworkspace.WorkspaceTreeItem): Promise<void> {
+	public async publishToDockerContainer(context: Project | dataworkspace.WorkspaceTreeItem): Promise<void> {
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/controllers/projectController.ts:262]: Update doc comment

File: extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts [hunk 3]
```diff
@@ -128,15 +61,69 @@ export async function launchDeployDatabaseQuickpick(project: Project): Promise<I
...
 
+	return {
+		envVariableName: envVarName,
+		appSettingFile: settingExist ? localSettings : undefined,
+		appSettingType: settingExist ? AppSettingType.AzureFunction : AppSettingType.None
+	};
+}
+
+/**
+ * Create flow for Deploying a database using only VS Code-native APIs such as QuickPick
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts:72]: Update to indicate this specifically publishes to docker

File: extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts [hunk 4]
```diff
@@ -128,15 +61,69 @@ export async function launchDeployDatabaseQuickpick(project: Project): Promise<I
...
+ */
+export async function launchPublishToDockerContainerQuickpick(project: Project): Promise<IDeployProfile | undefined> {
+
+	let localDbSetting: ILocalDbSetting | undefined;
+	// Deploy to docker selected
+	let portNumber = await vscode.window.showInputBox({
+		title: constants.enterPortNumber,
+		ignoreFocusOut: true,
+		value: constants.defaultPortNumber,
+		validateInput: input => isNaN(+input) ? constants.portMustBeNumber : undefined
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts:82]: Would check it's between 0 and 65,535 as well

File: extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts [hunk 5]
```diff
@@ -128,15 +61,69 @@ export async function launchDeployDatabaseQuickpick(project: Project): Promise<I
...
+	if (!portNumber) {
+		return undefined;
+	}
+
+	let password: string | undefined = '';
+	password = await vscode.window.showInputBox({
+		title: constants.enterPassword,
+		ignoreFocusOut: true,
+		value: password,
+		password: true
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts:96]: Should validate non-empty

File: extensions/sql-database-projects/src/dialogs/publishDatabaseQuickpick.ts [hunk 6]
```diff
@@ -207,17 +207,37 @@ export async function getPublishDatabaseSettings(project: Project, promptForConn
...
+			if (!action) {
+				return;
+			}
+			await projectController.publishOrScriptProject(project, settings, action === constants.publish);
 		}
-		await projectController.publishOrScriptProject(project, settings, action === constants.publish);
 	}
 }
 
+export async function lunchPublishTargetOption(): Promise<string | undefined> {
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/dialogs/publishDatabaseQuickpick.ts:229]: `launchPublishTargetOption`

Does this need to be exported?

File: extensions/sql-database-projects/src/common/constants.ts [hunk 7]
```diff
@@ -128,13 +128,13 @@ export const done = localize('done', "Done");
 export const nameMustNotBeEmpty = localize('nameMustNotBeEmpty', "Name must not be empty");
 
 // Deploy
-export const selectDeployOption = localize('selectDeployOption', "Select where to deploy the project to");
-export const deployToExistingServer = localize('deployToExistingServer', "Deploy to existing server");
-export const deployToDockerContainer = localize('deployToDockerContainer', "Deploy to docker container");
+export const selectDeployOption = localize('selectDeployOption', "Select where to publish the project to");
```
Comment (Charles-Gagnon) [inline extensions/sql-database-projects/src/common/constants.ts:131]: update ids and var names (`selectPublishOption` and such)

**Review by Charles-Gagnon — APPROVED** [review]

File: extensions/sql-database-projects/src/dialogs/publishDatabaseQuickpick.ts [hunk 8]
```diff
@@ -207,17 +207,37 @@ export async function getPublishDatabaseSettings(project: Project, promptForConn
...
-	let settings: IDeploySettings | undefined = await getPublishDatabaseSettings(project);
-
-	if (settings) {
-		// 5. Select action to take
-		const action = await vscode.window.showQuickPick(
-			[constants.generateScriptButtonText, constants.publish],
-			{ title: constants.chooseAction, ignoreFocusOut: true });
-		if (!action) {
-			return;
+	const publishTarget = await lunchPublishTargetOption();
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/publishDatabaseQuickpick.ts:210]: typo "lunch"

File: extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts [hunk 9]
```diff
@@ -5,83 +5,16 @@
...
-	let deploySettings = await getPublishDatabaseSettings(project, deployOption !== constants.deployToDockerContainer);
-
-	// Return when user hits escape
-	if (!deploySettings) {
-		return undefined;
-	}
-
-	// TODO: Ask for SQL CMD Variables or profile
-
+export async function launchDeployAppIntegrationQuickpick(project: Project): Promise<IDeployAppIntegrationProfile | undefined> {
```
Comment (kisantia) [inline extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts:17]: update file name to match

**Review by kisantia — APPROVED** [review]

File: extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts (hunk 9 above)
Comment (llali) [reply extensions/sql-database-projects/src/dialogs/deployDatabaseQuickpick.ts:17]: I didn't change the file name to keep all the actions for deploy (deploy workspace) here as well. deploy workspace will publish db project and deploy the app

**Pull request merged**